### PR TITLE
Add tudaexercise template

### PR DESCRIPTION
**Pull request by FussballAndy**
### Missing features

- [x] Proper coloring, change title text color based on accent color etc.
- [x] Add colorback option to make title not have color background
- [x] Add documentation

### Optionals

- [x] Darkmode. Similar to Rubens Latex template

### CODE REVIEW

File: templates/tudaexercise/template/tudaexercise.typ [hunk 1]
```diff
@@ -1,238 +1,149 @@
...
+      tuda-section(sec_title + " " + it.body)
+    } else if it.level == 2 {
+      let sec_title = ""
+      if info.sheetnumber != none {
+        sec_title += str(info.sheetnumber) + "."
+      }
+      sec_title += numbering("1a)", c.at(0), c.at(1))
+      tuda-subsection(sec_title + " " + it.body)
+    } else {
+      it
```
Comment (JeyRunner) [inline templates/tudaexercise/template/tudaexercise.typ:94]: I would define more of the numbering the actual numbering definition, like
```
set heading(
    numbering: (..numbers) => {
       prefix = ""
       if info.sheetnumber != none {
          prefix += str(info.sheetnumber) + "."
        }
      [#prefix]
      numbering("1 a", ..numbers)
    }
  )
```
This way the the chapters in the metadata (index) of the pdf are correct.
Then for the heading roule:
```
show heading: it => {
  let prefix = ""
  if it.level == 1 {
     prefix = dict.task + " "
   }
  let counter-suffix = if it.level > 1 {")"} else {":"}
  
  tuda-subsection({
       prefix
        counter(heading).display(it.numbering)
        counter-suffix
       it.body
   })
}
```

Comment (JeyRunner) [reply templates/tudaexercise/template/tudaexercise.typ:94]: Better to also have a check if the user disabled the numbering for this section, like here https://github.com/JeyRunner/tuda-typst-templates/blob/09e122ff52702439466b02c55f67d8430e6a7b97/common/headings.typ#L31

File: templates/tudaexercise/template/tudaexercise.typ [hunk 2]
```diff
@@ -1,238 +1,149 @@
...
+    rows: auto,
+    row-gutter: 1.4mm + 0.25mm,
+    identbar,
+    line(length: 100%, stroke: title_rule),
   )
 
-  // disable heading outlined for outline
-  set heading(outlined: false)
+  let inner_page_margin_top = 22pt
+  let logo_height = 22mm
```
Comment (JeyRunner) [inline templates/tudaexercise/template/tudaexercise.typ:114]: We should also better put this into props.typ since is is always the same

**Comment by JeyRunner** [comment]
Nice, it looks good.
I added some review comments.

**Comment by FussballAndy** [comment]
Most review points should be fixed now. I also added a few more utilities for coloring and working with dicts/default values.
One thing I am unsure about is whether we should give the user more customization options for coloring, so letting the user also be able to overwrite the background/text color from within the template. I am also thinking about adding a more general design state in which all design options like colors get stored. This will mostly just be used for the sections but will also give the user a way to access the colors used by the template.

File: templates/tudaexercise/template/sections.typ [hunk 3]
```diff
@@ -0,0 +1,33 @@
+// For now hardcoded value - should be half title rule height. Maybe change to state in the future however that will add a lot of computation complexity
+#let tuda_section_line_height = 0.6pt
+
+#let tuda-section-lines(above, below, body) = {
```
[the remark this answers is not in the document]
Comment (FussballAndy) [reply templates/tudaexercise/template/sections.typ]: I put the section functions into headings.typ. For now I also left tud-heading-with-lines as I am not sure if we should leave it for future use or not.

**Comment by JeyRunner** [comment]
Looks pretty nice :+1: 
Regarding the customization options, you can add custom background/text colouring if you want, but I am not sure if it is really required. I would generally go for: simpler = better. But if it is simple to implement its fine I guess.
In general, I would avoid using state if it is not needed. But your use case of giving the user access to computed colours makes sense (e.g. to make figures that match the accent colour). I would propose to make this design state just one way so that the user can't accidentally mess up the design by overwriting the state (and this way we effectively don't have state as long as the user does not read it). So essentially passing all computed colours to the sub-functions by value, but also writing the colours to the separate design state (but never read from it). What do you think?

Btw, you can also add your name to the typst.toml, so that you will be displayed as an author on the Universe website too.

**Comment by FussballAndy** [comment]
This should be mostly it. Though I am unsure if we should use the same README.md for both templates. Especially hence the images in these READMEs will also break in the typst packages repository. But also because (at least the current README) is rather specific to the thesis template.

**Comment by JeyRunner** [comment]
Yes looks pretty good.
Regarding the readme, I think with a few adaptations we could make the readme more general. 
e.g. in 'Usage' change to:
## Usage
Create a new typst project based on this template locally.
```bash
# for the thesis template
typst init @preview/athena-tu-darmstadt-thesis
cd athena-tu-darmstadt-thesis

# or for the exercise template
typst init @preview/athena-tu-darmstadt-exercise
cd athena-tu-darmstadt-exercise
```

<br><br>
There is already some functionality in the publish script to fix links which could be extended to all file links (just add the full path to this repo to all links).

But we could also have separate readmes per template. But I think most info is the same for all templates (disclaimer, font, logo setup).

**Comment by FussballAndy** [comment]
Yeah that sounds good. I did some changes to the README already. The script seems to only replace Markdown links and the images are embedded via img tags so I assume that is why that does not work right now. But that could just be done by hand for publishing. I will probably also remove some sections like TODOs and slightly change the `main.typ` setup but yeah for the rest the current README should be sufficient.

So yeah from my end everything is done so unless you have any more wishes or complaints I'd be open to this getting merged.

**Review by mohe2015 — APPROVED** [review]

**Comment by JeyRunner** [comment]
Whatever you prefer. But I can also merge it now and then you could open a new pull request for the following changes.

**Comment by JeyRunner** [comment]
I added fixing of image links to the publish script (just pull the changes from the main branch). It should work now automatically.

**Comment by FussballAndy** [comment]
Yeah then I'll probably submit it to universe first, doesn't really seem necessary to create a new PR for changes related to the first release.

**Comment by FussballAndy** [comment]
I created a PR at https://github.com/typst/packages/pull/1314. Please also have a look at it.

**Comment by FussballAndy** [comment]
The typst package check complains about the values in `props.typ` not being kebab case. I left it in there from your setup for the template. But now I'm unsure if we even need to make it publicly available. So should we perhaps remove it from `lib.typ`, change the variables to kebab-case or leave it?

**Comment by JeyRunner** [comment]
I would definitely keep it in the props file. Not having kebab case is generally not required for packages to be published. The thesis package got accepted this way.
So either change it to kebab or keep it as is.
I would keep it as is for now...

**Comment by FussballAndy** [comment]
The package just got merged. I'll add a few more changes I made to the README later, then this can also get merged.

**Comment by FussballAndy** [comment]
Alright should now be done. 👍

**Comment by JeyRunner** [comment]
Thanks for your great work @FussballAndy 🎉!
And congratulations on the accepted package.
I will merge it now.

**Pull request merged**